src/main/java/com/fasterxml/jackson/databind/util/clhm/ConcurrentLinkedHashMap.java

```diff
@@ -0,0 +1,1597 @@
+ * does <em>not</em> allow <tt>null</tt> to be used as a key or value. Unlike
+ * {@link java.util.LinkedHashMap}, this class does <em>not</em> provide
+ * predictable iteration order. A snapshot of the keys and entries may be
+ * obtained in ascending and descending order of retention.
+ *
+ * @author ben.manes@gmail.com (Ben Manes)
+ * @param <K> the type of keys maintained by this map
+ * @param <V> the type of mapped values
+ * @see <a href="http://code.google.com/p/concurrentlinkedhashmap/">
+ *      http://code.google.com/p/concurrentlinkedhashmap/</a>
+ */
+//@ThreadSafe
+public final class ConcurrentLinkedHashMap<K, V> extends AbstractMap<K, V>
+        implements ConcurrentMap<K, V>, Serializable {
+
+    /*
+     * This class performs a best-effort bounding of a ConcurrentHashMap using a
+     * page-replacement algorithm to determine which entries to evict when the
+     * capacity is exceeded.
+     *
+     * The page replacement algorithm's data structures are kept eventually
+     * consistent with the map. An update to the map and recording of reads may
+     * not be immediately reflected on the algorithm's data structures. These
+     * structures are guarded by a lock and operations are applied in batches to
+     * avoid lock contention. The penalty of applying the batches is spread across
+     * threads so that the amortized cost is slightly higher than performing just
+     * the ConcurrentHashMap operation.
+     *
+     * A memento of the reads and writes that were performed on the map are
+     * recorded in buffers. These buffers are drained at the first opportunity
+     * after a write or when the read buffer exceeds a threshold size. The reads
+     * are recorded in a lossy buffer, allowing the reordering operations to be
+     * discarded if the draining process cannot keep up. Due to the concurrent
+     * nature of the read and write operations a strict policy ordering is not
+     * possible, but is observably strict when single threaded.
+     *
+     * Due to a lack of a strict ordering guarantee, a task can be executed
+     * out-of-order, such as a removal followed by its addition. The state of the
+     * entry is encoded within the value's weight.
+     *
+     * Alive: The entry is in both the hash-table and the page replacement policy.
+     * This is represented by a positive weight.
+     *
+     * Retired: The entry is not in the hash-table and is pending removal from the
+     * page replacement policy. This is represented by a negative weight.
+     *
+     * Dead: The entry is not in the hash-table and is not in the page replacement
+     * policy. This is represented by a weight of zero.
+     *
+     * The Least Recently Used page replacement algorithm was chosen due to its
+     * simplicity, high hit rate, and ability to be implemented with O(1) time
+     * complexity.
+     */
+
+    /** The number of CPUs */
+    static final int NCPU = Runtime.getRuntime().availableProcessors();
```

Comment:
Do we want to do this, or just default to some static value? Due to embedded use, would want to
avoid access to `Runtime` I think.

Reply:
The jdk uses it, e.g. Striped64 (LongAdder) for sizing. I don’t recall any security manager, etc issues using that, but I could be wrong.

Reply:
Ok. I guess it's more a philosophical, not practical objection from my side.

Reply:
In Caffeine this is dynamically sized based on contention up to the cpu count. Here it’s fixed as earlier, low cpus were more common, and it was more specialized users. I think it would be fine if you prefer to set a conservative value.